### PR TITLE
*: Fix use of user identity in SHOW GRANTS + error messages

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Issue Number: fixes https://github.com/pingcap/tidb/issues/30279 (once new tidb-test PR merged as well)

Problem Summary:

https://github.com/pingcap/tidb/pull/30204 revealed a bug (i.e. it's not specifically a regression introduced) where the column used in `show grants` is from the Hostname of the user that logged in, not the matching entry from the MySQL privileges table.

Upon further inspection, it looks like we almost always intend to use the user from the privileges table and not the hostname of the logged in user. The exception is the `USER()` sql function.

(MySQL is actually inconsistent in which hostname it uses in error messages. The important case is with `SHOW GRANTS FOR CURRENT_USER()`  which is used in integration tests.)

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

```release-note
TiDB now more consistently prints the user+hostname of a user in error messages.
```

### CODE REVIEW

File: privilege/privileges/privileges_test.go [hunk 1]
```diff
@@ -372,6 +372,35 @@ func TestShowGrants(t *testing.T) {
 	require.Len(t, gs, 3)
 }
 
+// TestErrorMessage checks that the identity in error messages matches the mysql.user table one.
+// MySQL is inconsistent in its error messages, as some match the loginHost and others the
+// identity from mysql.user. In TiDB we now use the identity from mysql.user in error messages
+// for consistency.
+func TestErrorMessage(t *testing.T) {
```
Comment (dveeden) [inline privilege/privileges/privileges_test.go:379]: Manually testing this with v5.3.0 and master @ 13d0deb7fef7fd7391dce19f320abef4256223bd show the same error messages as in this test. Are you sure this test fails without your fix?

Comment (morgo) [reply privilege/privileges/privileges_test.go:379]: I added this for future checking.

The case where it currently fails prior to this is the Column name in `SHOW GRANTS FOR CURRENT_USER()`. But in the tests I didn't find an example where we check column names.

File: parser/auth/auth.go [hunk 2]
```diff
@@ -41,18 +41,24 @@ func (user *UserIdentity) Restore(ctx *format.RestoreCtx) error {
 }
 
 // String converts UserIdentity to the format user@host.
+// It defaults to providing the AuthIdentity (the matching entry in priv tables)
```
Comment (tiancaiamao) [inline parser/auth/auth.go:44]: NOTE: This totally change the semantic of the old API, and all the code referencing of the old API might be affected.

```
AuthIdentityString() -> String()
String() -> LoginString()
```

I can't say the new API is better or necessay (Go also discourage public API change), but as long as it fix more cases and do not introduce regression, it looks good to me.

**Review by tiancaiamao — APPROVED** [review]

**Review by dveeden — APPROVED** [review]

**Pull request merged**